Lombiq.NodeJs.Extensions/scripts/copy-assets.js

```diff
@@ -0,0 +1,104 @@
+function log(message) {
+    if (verbose) process.stdout.write(message + '\n');
+}
+
+async function copyFilesFromConfig(assetsConfig) {
+    return Promise.all(assetsConfig
+        .map((assetsGroup) => assetsGroup.sources.map((assetSource) => {
+            // Normalize the relative path to the directory to remove trailing slashes and straighten out any anomalies.
+            const directoryToCopy = path.relative('.', assetSource);
+            log(`directoryToCopy = ${directoryToCopy}`);
+
+            return access(directoryToCopy)
+                .catch(() => process.stderr.write(`The directory "${directoryToCopy}" cannot be accessed to copy files from.`))
+                .then(() => {
+                    const pathPattern = path.join(directoryToCopy, assetsGroup.pattern || filePattern);
+                    const sourceAndTargetPaths = [pathPattern, assetsGroup.target];
+
+                    // We want to copy all files matched by the given pattern into the target folder mirroring the
+                    // source folder structure. This is done by removing the source folder path from the beginning which
+                    // "copyfiles" does using the "up" option.
+                    const depth = directoryToCopy.split(/[\\/]/).length;
+
+                    // https://github.com/calvinmetcalf/copyfiles#programic-api; promisified using Node.utils.promisify.
+                    return copyfiles(sourceAndTargetPaths, { verbose: verbose, up: depth }, () => {});
+                });
+        }))
+        .reduce((previousArray, currentArray) => [...previousArray, ...currentArray], []));
+}
+
+async function readConfig(assetsFilePath) {
+    log('Searching configuration...');
+
+    return readFile(assetsFilePath, 'utf8')
+        .then((assetsConfig) => {
+            log(`Reading configuration from ${assetsFileName}...`);
+            return JSON.parse(assetsConfig);
+        })
+        .catch(() => readFile('package.json', 'utf8')
+            .then((packageConfig) => {
+                log('Reading configuration from package.json...');
+                return JSON.parse(packageConfig)[assetsKeyInPackageJson];
+            })
+        );
+}
+
+(async function main() {
+    log('Executing copy-assets.js...');
+
+    try {
+        await readConfig(assetsFileName)
+            .then((config) => {
+                if (config) {
+                    log(`Loaded assets config: ${JSON.stringify(config)}`);
+                    return copyFilesFromConfig(config);
+                }
+                log('No configuration found.');
+                return Promise.resolve();
+            })
+            .then(() => { log('Finished executing copy-assets.js. (1)'); })
+            .catch((inner) => process.stderr.write(JSON.stringify(inner)));
+
+        log('Finished executing copy-assets.js. (2)');
+    }
+    catch (error) {
+        process.stderr.write(JSON.stringify(error));
+    }
+
+    log('Finished executing copy-assets.js. (3)');
```

Comment:
@DAud-IcI Can you spot why neither of the three log statements are displayed while those from line 88 and those from inside the `copyFilesFromConfig()` call in line 89 are? It just doesn't make any sense to me.

Also, if I inline the `main()` method, and after removal of the `await` in line 85, the log statements `(2)` and `(3)` are displayed, but still not `(1)`. What gives?

Reply:
I've just checked this out and they get displayed for me:
![image](https://user-images.githubusercontent.com/4007293/181011025-f843260b-1748-4b1b-a31c-dd3623c76d1c.png)

> Also, if I inline the `main()` method, and after removal of the `await` in line 85, the log statements `(2)` and `(3)` are displayed, but still not `(1)`. What gives?

If you remove the `await` on line 85, then line 97 and onwards are not sequentially after the Promise any more. Since Javascript is strictly single-threaded, the way this works is that the Promise-returning method (`readConfig`) schedules an event (put it into an event queue) and then the scheduler will look at it once the current function scope is over. So at best you get (2), (3), (1). Also I'm not certain Node would wait for the Promise in this scenario since it wasn't returned.

Reply:
Thanks a lot for trying it out, Dávid! I'm perplex to see the output appeared for you.

I've noticed, though, that the line that's displayed just before (1) is "No configuration found" which might just be the explanation. Looks like `return copyFilesFromConfig()` in line 89 returns something that breaks that 😕 I'll have another look.

Reply:
After some more digging I now see the messages, too. 🎉